appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java

```diff
@@ -0,0 +1,696 @@
+
+        // Parse the Managed Beans
+        parseManagedBeans();
+
+        // Parse AppScoped resources
+        String appName = DOLUtils.getApplicationName(application);
+        Map<String, List<String>> resourcesList =
+                (Map<String, List<String>>) dc.getTransientAppMetadata().get(ResourceConstants.APP_SCOPED_RESOURCES_JNDI_NAMES);
+        myNamespace.storeAppScopedResources(resourcesList, appName);
+    }
+
+    private void parseManagedBeans() {
+        for (BundleDescriptor bd : application.getBundleDescriptors()) {
+            for (ManagedBeanDescriptor managedBean : bd.getManagedBeans()) {
+                myNamespace.store(managedBean.getGlobalJndiName(), (JndiNameEnvironment)bd);
+            }
+        }
+    }
+
+    private void parseResources(JndiNameEnvironment env) {
+        for (Object next : env.getResourceReferenceDescriptors()) {
+            parseResources((ResourceReferenceDescriptor) next, env);
+        }
+
+        for (Object next : env.getResourceEnvReferenceDescriptors()) {
+            parseResources((ResourceEnvReferenceDescriptor) next, env);
+        }
+
+        for (Object next : env.getMessageDestinationReferenceDescriptors()) {
+            storeResource(((MessageDestinationReferenceDescriptor) next).getName(), env);
+        }
+
+        for (Object next : env.getEnvironmentProperties()) {
+            storeResource(((EnvironmentProperty) next).getName(), env);
+        }
+
+        /**
+         * TODO: App client doesn't support some resource descriptors. Might need to fail deployment in such cases.
+         */
+        for (Object next : env.getAllResourcesDescriptors()) {
+            storeResource(((ResourceDescriptor) next).getName(), env);
+        }
+
+        for (Object next : env.getEntityManagerReferenceDescriptors()) {
+            storeResource(((EntityManagerReferenceDescriptor) next).getName(), env);
+        }
+
+        for (Object next : env.getEntityManagerFactoryReferenceDescriptors()) {
+            storeResource(((EntityManagerFactoryReferenceDescriptor) next).getName(), env);
+        }
+
+        for (Object next : env.getEjbReferenceDescriptors()) {
+            storeResource(((EjbReferenceDescriptor) next).getName(), env);
+        }
+
+        for (Object next : env.getServiceReferenceDescriptors()) {
+            storeResource(((ServiceReferenceDescriptor) next).getName(), env);
+        }
+    }
+
+    private void parseResources(ResourceReferenceDescriptor resRef, JndiNameEnvironment env) {
+        resRef.checkType();
+        storeResource(resRef.getName(), env);
+    }
+
+    private void parseResources(ResourceEnvReferenceDescriptor resEnvRef, JndiNameEnvironment env) {
+        resEnvRef.checkType();
+        storeResource(resEnvRef.getName(), env);
+    }
+
+    private void storeResource(String name, JndiNameEnvironment env) {
+        String logicalJNDIName = getLogicalJNDIName(name, env);
+        myNamespace.store(logicalJNDIName, env);
```

Comment:
I see that `myNamespace` is being used in various methods. If we are creating a `ResourceValidator` instance each time an application is being deployed, I guess this should be fine, else if we create only one such instance at server start, we would be better off passing it as a method argument.